Source/Request.swift

```diff
@@ -346,18 +340,30 @@ open class Request {
         retryOrFinish(error: self.error)
     }
 
-    /// Called when the `RequestDelegate` is retrying this `Request`.
+    /// Called when the `RequestDelegate` is retrying this `Request`. Calls `reset()`.
     func requestIsRetrying() {
         protectedMutableState.write { $0.retryCount += 1 }
 
+        reset()
+
         eventMonitor?.requestIsRetrying(self)
     }
 
     /// Called to trigger retry or finish this `Request`.
     func retryOrFinish(error: Error?) {
-        if let error = error, delegate?.willAttemptToRetryRequest(self) == true {
-            delegate?.retryRequest(self, ifNecessaryWithError: error)
-            return
+        if let error = error, let delegate = delegate {
+            delegate.retryResult(for: self, dueTo: error) { retryResult in
+                guard let retryResult = retryResult else { self.finish(); return }
+
+                switch retryResult {
+                case .doNotRetry:
+                    self.finish()
+                case .doNotRetryWithError(let error):
+                    self.finish(error: error)
+                default:
+                    delegate.retryRequest(self, withDelay: retryResult.delay)
+                }
```

Comment:
The act of retrying is now explicit by the caller rather than having to know that the delegate has already triggered the retry. Much better design overall.

> As a side note, it seems weird to me that we have an optional delegate. Nothing works if it isn't set, so I wonder if we can rethink the memory management (maybe even the naming), and make it a strong, non-optional property. Something to consider in another PR.

Reply:
We can discuss it more, but in general, we need a way to break the reference cycle between the `Session` and its `Request`s, and the `weak` delegate does that. We could treat it as `unowned`, but we can't guarantee lifetimes like that. If we do want a strong delegate, both a rename / redesign of the API is in order, as well as another solution to the cycle issue.

Reply:
Absolutely, we certainly don't want to introduce a reference cycle. My thought was more around your final point of just rethinking the API to be less of a delegate and more of a core set of functionality that's required by the `Request` type.